File: dashboard/workbench_es.go
package main

// workbench_es.go — #405 follow-up: Payload Workbench recipe/run storage,
// Elasticsearch-backed. See workbench_domain.go's package comment on
// workbenchService for the full design rationale (why this was previously
// local-disk-only, and why a run's document ID being its own idempotency
// key makes atomic op_type=create the right primitive here). No local
// fallback: every method below returns an error when es is nil, same
// posture as #494/#483/#475.

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

const workbenchRunsIndex = "dashboard-workbench-runs-v1"
const workbenchRecipesIndex = "dashboard-workbench-recipes-v1"

// workbenchUpdateRetries bounds updateRun's read-modify-write retry loop
// against a losing compare-and-swap -- the same shape and reasoning as
// alertManager's alertWriteRetries (alerts.go).
const workbenchUpdateRetries = 5

func workbenchRecipeDocID(id string, revision int) string {
	return id + ":" + strconv.Itoa(revision)
}

// listRecipes returns every recipe visible to owner (their own private
// recipes plus every shared one), newest-then-highest-revision first.
func (w *workbenchService) listRecipes(owner string) []workbenchRecipe {
	if !w.configured() {
		return nil
	}
	hits, err := w.es.docSearchAll(workbenchRecipesIndex, 10000)
	if err != nil {
		return nil
	}
	visible := make([]workbenchRecipe, 0, len(hits))
	for _, hit := range hits {
		var recipe workbenchRecipe
		if json.Unmarshal(hit.Source, &recipe) != nil {
			continue
		}
		if recipe.SchemaVersion != workbenchSchemaVersion || !validWorkbenchID(recipe.ID, "recipe_") || recipe.Revision <= 0 || recipe.Owner == "" {
			continue
		}
		if recipe.Scope == "shared" || recipe.Owner == owner {
			visible = append(visible, recipe)
		}
	}
	sort.Slice(visible, func(i, j int) bool {
		if visible[i].CreatedAt.Equal(visible[j].CreatedAt) {
			return visible[i].Revision > visible[j].Revision
		}
		return visible[i].CreatedAt.After(visible[j].CreatedAt)
	})
	return visible
}

// saveRecipe creates a new recipe (input.ID == "") or a new revision of an
// existing one, enforcing the same optimistic-concurrency contract the
// local-file version did (a stale baseRevision returns errWorkbenchConflict
// rather than silently overwriting a concurrent save) -- except the
// compare-and-swap is now Elasticsearch's own op_type=create on the
// revision-numbered document (workbenchRecipeDocID), not a whole-file
// rewrite guarded by a mutex.
func (w *workbenchService) saveRecipe(input workbenchRecipe, owner string, baseRevision int) (workbenchRecipe, error) {
	if !w.configured() {
		return workbenchRecipe{}, errors.New("workbench persistence is not configured")
	}
	input.Name = strings.TrimSpace(input.Name)
	input.Description = strings.TrimSpace(input.Description)
	if len(input.Name) < 2 || len(input.Name) > 80 || len(input.Description) > 400 {
		return workbenchRecipe{}, errors.New("recipe name or description is outside the allowed length")
	}
	if input.Scope != "private" && input.Scope != "shared" {
		return workbenchRecipe{}, errors.New("recipe scope must be private or shared")
	}
	selections, err := validateWorkbenchSelections(input.Analyzers)
	if err != nil {
		return workbenchRecipe{}, err
	}
	hits, err := w.es.docSearchAll(workbenchRecipesIndex, 10000)
	if err != nil {
		return workbenchRecipe{}, err
	}
	if input.ID == "" {
		// #883: workbenchMaxRecipes bounds how many distinct recipes exist,
		// not how many revision documents they've accumulated -- every save
		// (including an edit to an already-existing recipe) writes a new
		// workbenchRecipeDocID(id, revision) document that's never pruned,
		// so counting raw hits here let ordinary editing activity alone
		// permanently exhaust the cap. Only a genuinely new recipe ID needs
		// to check it; an edit of an existing recipe is handled below.
		ids := make(map[string]bool, len(hits))
		for _, hit := range hits {
			var recipe workbenchRecipe
			if json.Unmarshal(hit.Source, &recipe) != nil {
				continue
			}
			ids[recipe.ID] = true
		}
		if len(ids) >= workbenchMaxRecipes {
			return workbenchRecipe{}, errors.New("recipe limit reached")
		}
		input.ID, err = randomWorkbenchID("recipe")
		if err != nil {
			return workbenchRecipe{}, err
		}
		input.Revision = 1
	} else {
		if !validWorkbenchID(input.ID, "recipe_") {
			return workbenchRecipe{}, errors.New("invalid recipe id")
		}
		latest := 0
		for _, hit := range hits {
			var recipe workbenchRecipe
			if json.Unmarshal(hit.Source, &recipe) != nil || recipe.ID != input.ID {
				continue
			}
			if recipe.Owner != owner {
				return workbenchRecipe{}, errWorkbenchNotFound
			}
			latest = max(latest, recipe.Revision)
		}
		if latest == 0 {
			return workbenchRecipe{}, errWorkbenchNotFound
		}
		if baseRevision != latest {
			return workbenchRecipe{}, errWorkbenchConflict
		}
		input.Revision = latest + 1
	}
	input.SchemaVersion = workbenchSchemaVersion
	input.Owner = owner
	input.CreatedAt = time.Now().UTC()
	input.Analyzers = selections
	body, err := json.Marshal(input)
	if err != nil {
		return workbenchRecipe{}, err
	}
	// create=true: this exact revision number is written exactly once --
	// two racing saves both computing "latest+1" from the same read collide
	// here instead of one silently clobbering the other.
	if err := w.es.docIndex(workbenchRecipesIndex, workbenchRecipeDocID(input.ID, input.Revision), body, true, 0, 0); err != nil {
		if errors.Is(err, errESConflict) {
			return workbenchRecipe{}, errWorkbenchConflict
		}
		return workbenchRecipe{}, err
	}
	return input, nil
}

func (w *workbenchService) recipe(id string, revision int, owner string) (workbenchRecipe, error) {
	if !w.configured() {
		return workbenchRecipe{}, errWorkbenchNotFound
	}
	hit, found, err := w.es.docGet(workbenchRecipesIndex, workbenchRecipeDocID(id, revision))
	if err != nil || !found {
		return workbenchRecipe{}, errWorkbenchNotFound
	}
	var recipe workbenchRecipe
	if json.Unmarshal(hit.Source, &recipe) != nil {
		return workbenchRecipe{}, errWorkbenchNotFound
	}
	if recipe.Scope != "shared" && recipe.Owner != owner {
		return workbenchRecipe{}, errWorkbenchNotFound
	}
	return recipe, nil
}

// findRun fetches one run by id, scoped to owner -- the same cross-owner
// isolation the local-file version enforced (a run belonging to a different
// owner reports errWorkbenchNotFound, not a permission error, so its
// existence isn't leaked).
func (w *workbenchService) findRun(id, owner string) (workbenchRun, error) {
	if !w.configured() || !validWorkbenchRunID(id) {
		return workbenchRun{}, errWorkbenchNotFound
	}
	hit, found, err := w.es.docGet(workbenchRunsIndex, id)
	if err != nil || !found {
		return workbenchRun{}, errWorkbenchNotFound
	}
	var run workbenchRun
	if json.Unmarshal(hit.Source, &run) != nil || run.ID != id || run.Owner != owner {
		return workbenchRun{}, errWorkbenchNotFound
	}
	return run, nil
}

func (w *workbenchService) listRuns(hash, owner string) []workbenchRun {
	return w.listRunsForOwnerAndHash(owner, hash, 25)
}

func (w *workbenchService) listRunsForOwner(owner string, limit int) []workbenchRun {
	return w.listRunsForOwnerAndHash(owner, "", limit)
}

func (w *workbenchService) listRunsForOwnerAndHash(owner, hash string, limit int) []workbenchRun {
	if !w.configured() {
		return nil
	}
	if limit <= 0 || limit > workbenchMaxRuns {
		limit = workbenchMaxRuns
	}
	hits, err := w.es.docSearchAll(workbenchRunsIndex, 10000)
	if err != nil {
		return nil
	}
	all := make([]workbenchRun, 0, len(hits))
	for _, hit := range hits {
		var run workbenchRun
		if json.Unmarshal(hit.Source, &run) == nil && run.SchemaVersion == workbenchSchemaVersion && validWorkbenchRunID(run.ID) {
			all = append(all, run)
		}
	}
	sort.Slice(all, func(i, j int) bool { return all[i].CreatedAt.After(all[j].CreatedAt) })
	visible := make([]workbenchRun, 0, limit)
	for _, run := range all {
		if run.Owner == owner && (hash == "" || run.PayloadSHA256 == hash) {
			visible = append(visible, run)
			if len(visible) == limit {
				break
			}
		}
	}
	return visible
}

// countRuns backs createWorkbenchRun's retention check (reject a new run
// past workbenchMaxRuns rather than silently pruning old ones -- an
// operator archiving/investigating old runs must not have them vanish
// underneath them, unlike #475's generated-report retention which is
// disposable by design).
func (w *workbenchService) countRuns() (int, error) {
	if !w.configured() {
		return 0, errors.New("workbench persistence is not configured")
	}
	hits, err := w.es.docSearchAll(workbenchRunsIndex, 10000)
	if err != nil {
		return 0, err
	}
	return len(hits), nil
}

// createOrReuseRun is the atomic half of createWorkbenchRun: run.ID is
// already the request's own idempotency key (see workbenchIdempotency), so
// op_type=create either wins (a genuinely new run) or collides with an
// identical prior submission -- in which case the existing run is fetched
// and returned as reused=true, the same "duplicate submission returns the
// original" contract the old in-process mutex enforced, now safe across
// dashboard instances too since Elasticsearch's own create semantics are
// what's doing the compare-and-swap, not a process-local lock.
func (w *workbenchService) createOrReuseRun(run workbenchRun) (workbenchRun, bool, error) {
	if !w.configured() {
		return workbenchRun{}, false, errors.New("analysis workbench persistence is not configured")
	}
	body, err := json.Marshal(run)
	if err != nil {
		return workbenchRun{}, false, err
	}
	err = w.es.docIndex(workbenchRunsIndex, run.ID, body, true, 0, 0)
	if err == nil {
		return run, false, nil
	}
	if !errors.Is(err, errESConflict) {
		return workbenchRun{}, false, err
	}
	hit, found, getErr := w.es.docGet(workbenchRunsIndex, run.ID)
	if getErr != nil || !found {
		return workbenchRun{}, false, fmt.Errorf("workbench run creation conflict could not be resolved: %w", getErr)
	}
	var existing workbenchRun
	if json.Unmarshal(hit.Source, &existing) != nil {
		return workbenchRun{}, false, errors.New("existing workbench run is corrupt")
	}
	return existing, true, nil
}

// updateRun is the read-modify-write primitive every in-place run mutation
// (reconciliation on read, a child cancel/retry action) goes through:
// fetch, let mutate decide the new state and whether anything actually
// changed, and persist only if so -- preserving the local-file version's
// "skip the write entirely when reconciliation found nothing new" behavior
// (TestReconcileWorkbenchRunSkipsWorkOnceEveryChildIsTerminal), just via a
// conditional Elasticsearch write instead of comparing file mtimes. mutate
// returning an error aborts without writing and propagates that error
// directly (a child action's own validation failure -- "only a queued child
// can be cancelled" and friends -- not a storage problem). A lost
// compare-and-swap race retries against a fresh read, same shape as
// alertManager's own retry loop.
func (w *workbenchService) updateRun(id, owner string, mutate func(*workbenchRun) (bool, error)) (workbenchRun, error) {
	if !w.configured() {
		return workbenchRun{}, errors.New("workbench persistence is not configured")
	}
	if !validWorkbenchRunID(id) {
		return workbenchRun{}, errWorkbenchNotFound
	}
	for attempt := 0; attempt < workbenchUpdateRetries; attempt++ {
		hit, found, err := w.es.docGet(workbenchRunsIndex, id)
		if err != nil {
			return workbenchRun{}, err
		}
		if !found {
			return workbenchRun{}, errWorkbenchNotFound
		}
		var run workbenchRun
		if json.Unmarshal(hit.Source, &run) != nil || run.ID != id || run.Owner != owner {
			return workbenchRun{}, errWorkbenchNotFound
		}
		changed, mutateErr := mutate(&run)
		if mutateErr != nil {
			return workbenchRun{}, mutateErr
		}
		if !changed {
			return run, nil
		}
		body, err := json.Marshal(run)
		if err != nil {
			return workbenchRun{}, err
		}
		err = w.es.docIndex(workbenchRunsIndex, id, body, false, hit.SeqNo, hit.PrimaryTerm)
		if err == nil {
			return run, nil
		}
		if !errors.Is(err, errESConflict) {
			return workbenchRun{}, err
		}
		// Lost a race against a concurrent update (another reconcile-on-read,
		// or a child action) -- retry from a fresh read.
	}
	return workbenchRun{}, errors.New("workbench run update conflict; retry")
}

File: arcane/home/honeypot-dashboard/backend-service/src/workbench_es.rs
//! Payload Workbench recipe/run persistence, ported from dashboard/
//! workbench_es.go. A run's own document id IS its idempotency key
//! (workbench_domain::idempotency_key), so `es.index_doc_create` gives
//! atomic, cross-instance create-or-conflict for free — a losing race is a
//! genuine "someone already submitted this," not a retry case. Recipes get
//! the analogous treatment: each revision is its own document
//! (`{id}:{revision}`), also written via create, so two racing saves can't
//! both claim the same revision number.

use crate::es::{Es, WriteError};
use crate::workbench_domain::{WorkbenchRecipe, WorkbenchRun};

pub const RUNS_INDEX: &str = "dashboard-workbench-runs-v1";
pub const RECIPES_INDEX: &str = "dashboard-workbench-recipes-v1";
const UPDATE_RETRIES: u32 = 5;

pub fn recipe_doc_id(id: &str, revision: i64) -> String {
    format!("{id}:{revision}")
}

pub async fn list_recipes(es: &Es, owner: &str) -> anyhow::Result<Vec<WorkbenchRecipe>> {
    let result = es
        .search_index(&[RECIPES_INDEX], serde_json::json!({"size": 10000}))
        .await?;
    let mut visible: Vec<WorkbenchRecipe> = result["hits"]["hits"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|hit| serde_json::from_value::<WorkbenchRecipe>(hit["_source"].clone()).ok())
        .filter(|recipe| {
            recipe.schema_version == crate::workbench_domain::SCHEMA_VERSION
                && crate::workbench_domain::valid_id(&recipe.id, "recipe_")
                && recipe.revision > 0
                && !recipe.owner.is_empty()
        })
        .filter(|recipe| recipe.scope == "shared" || recipe.owner == owner)
        .collect();
    visible.sort_by(|a, b| match b.created_at.cmp(&a.created_at) {
        std::cmp::Ordering::Equal => b.revision.cmp(&a.revision),
        other => other,
    });
    Ok(visible)
}

pub async fn recipe(es: &Es, id: &str, revision: i64, owner: &str) -> Option<WorkbenchRecipe> {
    let doc = es
        .get_doc(RECIPES_INDEX, &recipe_doc_id(id, revision))
        .await
        .ok()
        .flatten()?;
    let recipe: WorkbenchRecipe = serde_json::from_value(doc).ok()?;
    if recipe.scope != "shared" && recipe.owner != owner {
        return None;
    }
    Some(recipe)
}

pub enum SaveRecipeError {
    Validation(String),
    Conflict,
    NotFound,
    Storage(anyhow::Error),
}

/// input.id == "" creates a new recipe (revision 1); otherwise saves a new
/// revision of an existing one owned by `owner`, gated on `base_revision`
/// matching the current latest (a stale base is Conflict, mirroring the
/// old file-backed store's optimistic-concurrency contract).
pub async fn save_recipe(
    es: &Es,
    mut input: WorkbenchRecipe,
    owner: &str,
    base_revision: i64,
) -> Result<WorkbenchRecipe, SaveRecipeError> {
    input.name = input.name.trim().to_string();
    input.description = input.description.trim().to_string();
    if input.name.len() < 2 || input.name.len() > 80 || input.description.len() > 400 {
        return Err(SaveRecipeError::Validation(
            "recipe name or description is outside the allowed length".into(),
        ));
    }
    if input.scope != "private" && input.scope != "shared" {
        return Err(SaveRecipeError::Validation(
            "recipe scope must be private or shared".into(),
        ));
    }
    input.analyzers = crate::workbench_domain::validate_selections(input.analyzers)
        .map_err(SaveRecipeError::Validation)?;

    let result = es
        .search_index(&[RECIPES_INDEX], serde_json::json!({"size": 10000}))
        .await
        .map_err(SaveRecipeError::Storage)?;
    let hits: Vec<WorkbenchRecipe> = result["hits"]["hits"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|hit| serde_json::from_value::<WorkbenchRecipe>(hit["_source"].clone()).ok())
        .collect();

    if input.id.is_empty() {
        let distinct: std::collections::HashSet<&str> =
            hits.iter().map(|r| r.id.as_str()).collect();
        if distinct.len() >= crate::workbench_domain::MAX_RECIPES {
            return Err(SaveRecipeError::Validation("recipe limit reached".into()));
        }
        input.id = crate::workbench_domain::random_id("recipe");
        input.revision = 1;
    } else {
        if !crate::workbench_domain::valid_id(&input.id, "recipe_") {
            return Err(SaveRecipeError::Validation("invalid recipe id".into()));
        }
        let mut latest = 0;
        for existing in &hits {
            if existing.id != input.id {
                continue;
            }
            if existing.owner != owner {
                return Err(SaveRecipeError::NotFound);
            }
            latest = latest.max(existing.revision);
        }
        if latest == 0 {
            return Err(SaveRecipeError::NotFound);
        }
        if base_revision != latest {
            return Err(SaveRecipeError::Conflict);
        }
        input.revision = latest + 1;
    }
    input.schema_version = crate::workbench_domain::SCHEMA_VERSION;
    input.owner = owner.to_string();
    input.created_at = chrono::Utc::now().to_rfc3339();

    let doc =
        serde_json::to_value(&input).map_err(|error| SaveRecipeError::Storage(error.into()))?;
    match es
        .index_doc_create(
            RECIPES_INDEX,
            &recipe_doc_id(&input.id, input.revision),
            doc,
        )
        .await
    {
        Ok(()) => Ok(input),
        Err(WriteError::Conflict) => Err(SaveRecipeError::Conflict),
        Err(WriteError::Other(error)) => Err(SaveRecipeError::Storage(error)),
    }
}

pub async fn find_run(es: &Es, id: &str, owner: &str) -> Option<WorkbenchRun> {
    if !crate::workbench_domain::valid_run_id(id) {
        return None;
    }
    let doc = es.get_doc(RUNS_INDEX, id).await.ok().flatten()?;
    let run: WorkbenchRun = serde_json::from_value(doc).ok()?;
    if run.id != id || run.owner != owner {
        return None;
    }
    Some(run)
}

pub async fn list_runs_for_owner_and_hash(
    es: &Es,
    owner: &str,
    hash: &str,
    limit: usize,
) -> anyhow::Result<Vec<WorkbenchRun>> {
    let limit = if limit == 0 || limit > crate::workbench_domain::MAX_RUNS {
        crate::workbench_domain::MAX_RUNS
    } else {
        limit
    };
    let result = es
        .search_index(&[RUNS_INDEX], serde_json::json!({"size": 10000}))
        .await?;
    let mut all: Vec<WorkbenchRun> = result["hits"]["hits"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|hit| serde_json::from_value::<WorkbenchRun>(hit["_source"].clone()).ok())
        .filter(|run| {
            run.schema_version == crate::workbench_domain::SCHEMA_VERSION
                && crate::workbench_domain::valid_run_id(&run.id)
        })
        .collect();
    all.sort_by(|a, b| b.created_at.cmp(&a.created_at));
    Ok(all
        .into_iter()
        .filter(|run| run.owner == owner && (hash.is_empty() || run.payload_sha256 == hash))
        .take(limit)
        .collect())
}

pub async fn count_runs(es: &Es) -> anyhow::Result<usize> {
    let result = es
        .search_index(&[RUNS_INDEX], serde_json::json!({"size": 10000}))
        .await?;
    Ok(result["hits"]["hits"]
        .as_array()
        .map(|hits| hits.len())
        .unwrap_or(0))
}

/// Atomic create-or-reuse: run.id is already the idempotency key, so
/// op_type=create either wins (genuinely new) or collides with an
/// identical prior submission — in which case the existing run is fetched
/// and returned as reused=true.
pub async fn create_or_reuse_run(
    es: &Es,
    run: WorkbenchRun,
) -> anyhow::Result<(WorkbenchRun, bool)> {
    let doc = serde_json::to_value(&run)?;
    match es.index_doc_create(RUNS_INDEX, &run.id, doc).await {
        Ok(()) => Ok((run, false)),
        Err(WriteError::Conflict) => {
            let existing = es.get_doc(RUNS_INDEX, &run.id).await?.ok_or_else(|| {
                anyhow::anyhow!("workbench run creation conflict could not be resolved")
            })?;
            let existing: WorkbenchRun = serde_json::from_value(existing)?;
            Ok((existing, true))
        }
        Err(WriteError::Other(error)) => Err(error),
    }
}

pub enum UpdateRunError {
    NotFound,
    Mutate(String),
    Storage(anyhow::Error),
}

/// Read-modify-write primitive every in-place run mutation goes through:
/// fetch, let `mutate` decide the new state and whether anything actually
/// changed, persist only if so. A lost compare-and-swap race retries
/// against a fresh read, same shape as the alert notifier's own retry loop.
pub async fn update_run<F>(
    es: &Es,
    id: &str,
    owner: &str,
    mut mutate: F,
) -> Result<WorkbenchRun, UpdateRunError>
where
    F: FnMut(&mut WorkbenchRun) -> Result<bool, String>,
{
    if !crate::workbench_domain::valid_run_id(id) {
        return Err(UpdateRunError::NotFound);
    }
    for _ in 0..UPDATE_RETRIES {
        let Some((doc, seq_no, primary_term)) = es
            .get_doc_meta(RUNS_INDEX, id)
            .await
            .map_err(UpdateRunError::Storage)?
        else {
            return Err(UpdateRunError::NotFound);
        };
        let mut run: WorkbenchRun =
            serde_json::from_value(doc).map_err(|error| UpdateRunError::Storage(error.into()))?;
        if run.id != id || run.owner != owner {
            return Err(UpdateRunError::NotFound);
        }
        let changed = mutate(&mut run).map_err(UpdateRunError::Mutate)?;
        if !changed {
            return Ok(run);
        }
        let body =
            serde_json::to_value(&run).map_err(|error| UpdateRunError::Storage(error.into()))?;
        match es
            .index_doc_cas(RUNS_INDEX, id, body, seq_no, primary_term)
            .await
        {
            Ok(()) => return Ok(run),
            Err(WriteError::Conflict) => continue,
            Err(WriteError::Other(error)) => return Err(UpdateRunError::Storage(error)),
        }
    }
    Err(UpdateRunError::Storage(anyhow::anyhow!(
        "workbench run update conflict; retry"
    )))
}
